### DISCUSSION TITLE
Help setting sensor state

### DISCUSSION THREAD

**Discussion by britkat1980** [GrandMoff100/HomeAssistantAPI#9] [Q&A] [answered]
Sorry to ask this here, but can't find any documentation to help me use this to set a sensor state. I have a simple single value payload and entitiy_id, but am not sure what the right function call is. example.py only shows light control.
I know its going to be really straightforward, but can't see how to do it!

**Comment by GrandMoff100** [comment]
First off I want to thank you for your interest in the project. You're right there isn't any documentation on it yet. Upon further investigation, I've found that there's an endpoint on home assistant that we haven't made a function call for yet. I assume you want to change the sensor state in Home assistant but don't want to trigger any real world events which would be done with service triggering. If this is what you're looking for then I apologize for not including it yet. I'll get right on it and will let you know when it's added! Thanks!

**Reply by GrandMoff100** [reply]
I forgot to tell you that it's perfectly fine to ask questions here. We even encourage it.

So, would something like this work?
```py
client.set_state("<entity_id>", "<state>", **attributes)
```

In practice would look more like this.
```py
client.set_state("sun.sun", "below_horizon", next_rising="2016-05-31T03:39:14+00:00")
```

If this isn't what you're looking for please feel free to tell us.

**Reply by britkat1980** [reply]
That looks perfect, thanks

**Comment by britkat1980** [comment]
Yes, I’ve got some code which talks via modbus to a solar invertor and I want to update states in HA with the stats like solar, load, export, grid etc… if there’s a way to simply do that that’d be awesome! If it can be done as a service then happy to do it that way, if you can show me how.

**Comment by GrandMoff100** [comment]
The biggest question I have which basically decides which function you need to use, is whether or not you need your entity states backed by real world hardware components.

**Reply by britkat1980** [reply]
No need for this to be backed up by real world changes. They are all essentially virtual inputs

**Comment by GrandMoff100** [comment]
Sounds great! I'll get working on adding that function to the library and I'll let you know when it's done!

**Comment by GrandMoff100** [comment]
Should only be a few hours to complete.

**Comment by GrandMoff100** [comment]
Okay. It's now added! To utilize the new function `set_entity` upgrade to the newest version 0.1.1 and modify the example below to fit the example below to fit your needs.
(I named the function set_entity to be consistent with get_entity)

```py
client.set_entity('my_fabulous.entity', state='fabulous_of_course', some_attribute='foobar')
```

Now as you noted there is a lack of documentation to the project. The next steps for us would be to probably start working on that and we have a few ideas. So stay tuned!

**Reply by britkat1980** [reply]
Thanks for this, I'm just trying it now. I'm getting an error when running this code:
```
    def push(data):
        url=GiV_Settings.HA_url
        token=GiV_Settings.HA_token
        client = Client(url, token)
        power_output = data['Power']
        GivTCP.debug("Creating Power entities for HA")
        for key in power_output:
            entityid="sensor.givtcp_"+key.replace(" ","_")
            payload='{"state":"'+str(power_output[key])+'"}'
            print("sending",str(power_output[key]), "to", entityid) 
            result=client.set_entity(entityid, state=str(power_output[key]))
            print (result)
```

Error processing input registers: (<class 'KeyError'>, KeyError('entity_id'), <traceback object at 0x7f5bddfa4a00>)

Any ideas? The entity doesn't exist yet, but I assumed that it would create it automatically, is this the problem?

**Reply by GrandMoff100** [reply]
Could you send the traceback of your error?

**Reply by britkat1980** [reply]
Here's the error output:
![image](https://user-images.githubusercontent.com/69121158/123602113-56019800-d7f0-11eb-844e-515edb444f04.png)

**Comment by GrandMoff100** [comment]
Try upgrading to the latest debug version "0.1.1.post1" and then running your code. It should print the json response from HA so we can verify that this is exactly where the KeyError is being raised from, like the error output says it is.

**Reply by britkat1980** [reply]
![image](https://user-images.githubusercontent.com/69121158/123656926-3dad6f80-d828-11eb-96c3-7016aebd6904.png)

**Reply by GrandMoff100** [reply]
Found it!

**Comment by GrandMoff100** [comment]
Ok after upgrading to version "0.1.2" The empty attributes key bug the json body of set_entity *should* be fixed! Let me know if it's not. Thanks!

**Reply by britkat1980** [reply]
Still no joy...
![image](https://user-images.githubusercontent.com/69121158/123663567-6df80c80-d82e-11eb-96da-5faf1bcdec25.png)

**Reply by britkat1980** [reply]
The code I'm using is below. If I call get_entities() I do get the long list, so I know the connect is working.
```
import json
import sys
from homeassistant_api import Client
from settings import GiV_Settings

url=GiV_Settings.HA_url
token=GiV_Settings.HA_token
client = Client(url, token)
result=client.set_entity('sensor.givenergy_load_power', state='9999')
print (result)
```

**Reply by GrandMoff100** [reply]
Well Client automatically checks connecting when it's initialized so calling get_entities doesn't verify anything extra. I think there's still something wrong with my JSON. This is gonna take a while to figure out or maybe not. I'll let you know if I figure anything new out.

**Reply by GrandMoff100** [reply]
I think I may have found something. I think it may be a problem with the lack of a Content-Length header in Post requests with a body, currently testing this idea.

**Comment by GrandMoff100** [comment]
Found it for real!

In "https://github.com/home-assistant/core/blob/dev/homeassistant/components/api/__init__.py" on line 254-257:
```py
try:
    data = await request.json()
except ValueError:
    return self.json_message("Invalid JSON specified.", HTTP_BAD_REQUEST)
```

This is where the error is coming from so now I think I know how to fix it! Expect solution soon.

**Comment by GrandMoff100** [comment, marked as answer]
After much research I have concluded that it is a bug where the Content-Length header integer isn't matching the length of the post json body, which is a problem with requests so there's not much I can do about that so I'm going to open an issue with all relevant information on both python requests GitHub repository and one of this one labeling it a bug. Stay tuned for updates on this, with the upcoming developments planned for this project I think we will get it fixed soon! Thanks for your patience.
